types/puppeteer/index.d.ts

```diff
@@ -275,13 +275,17 @@ export interface JSEvalable<A = any> {
      * The only difference between `evaluate` and `evaluateHandle` is that `evaluateHandle` returns in-page object (`JSHandle`).
      * If the function, passed to the `evaluateHandle`, returns a `Promise`, then `evaluateHandle` would wait for the
      * promise to resolve and return its value.
-     * @param fn Function to be evaluated in browser context
-     * @param args Arguments to pass to `fn`
-     */
-    evaluateHandle(
-        pageFunction: string | ((arg1: A, ...args: any[]) => any),
-        ...args: SerializableOrJSHandle[]
-    ): Promise<JSHandle>;
+     * The TypeScript definitions assume that `evaluateHandle` returns a `JSHandle`, but if you know it's going to return an
+     * `ElementHandle`, pass it as the generic argument:
+     * @param pageFunction - a function that is run within the page
+     * @param args - arguments to be passed to the pageFunction
+     */
+    // tslint:disable-next-line no-unnecessary-generics (This generic is meant to be passed explicitly.) # https://github.com/Microsoft/dtslint/issues/76
+    evaluateHandle<HandlerType extends JSHandle = JSHandle>(
+      pageFunction: string | ((arg1: A, ...args: any[]) => any),
+      ...args: SerializableOrJSHandle[]
+  // tslint:disable-next-line no-unnecessary-generics (This generic is meant to be passed explicitly.) # https://github.com/Microsoft/dtslint/issues/76
```

Comment:
I'm surprised that we need these `tslint:disable`s – do you know why it thinks they are unnecessary?

Reply:
I'm not confident either, but puppeteer's doc said:
```
This function will return a JSHandle by default, however if your pageFunction returns an HTML element you will get back an ElementHandle:
```
When I first met this situation, I thought when using <code>evaluateHandle</code> in typescript would become like this:
```
const elementHandle = await page.evaluateHandle<HTMLAnchorElement>( fn ) // and fn returns HTMLAnchorElement
```
But, puppeteer [source comment](https://github.com/puppeteer/puppeteer/blob/444c7e0e4729748d4c124bf8b4948a3d74dcf726/src/common/Page.ts#L768:) said we should use:
```
const button = await page.evaluateHandle<ElementHandle>(...);
```
So, I tried [same type definition](https://github.com/puppeteer/puppeteer/blob/444c7e0e4729748d4c124bf8b4948a3d74dcf726/src/common/Page.ts#L774) to that code but, the <code>HandlerType</code> generic used only return type and cause <code>no-unnecessary-generics</code>. <code>HandlerType</code> depends on <code>pageFunction</code>'s return type but it doesn't have any relation with <code>JSHandle</code> or <code>ElementHandle</code> type(or just I failed to make fancier solution).

I think this issue that discussed in https://github.com/Microsoft/dtslint/issues/76. So, I tried the solution in the link https://github.com/Microsoft/dtslint/issues/76#issuecomment-364197479

Reply:
While I have to admit I'm curious why these types are here when puppeteer also ships a copy, it seems like matching their version is the way to go:
```ts
evaluateHandle<HandlerType extends JSHandle = JSHandle>(pageFunction: EvaluateHandleFn, ...args: SerializableOrJSHandle[]): Promise<HandlerType>;
```

Reply:
(i.e. I'm okay with suppressing the lint warning)